File: stats.py
#############################################################
# Title: Summary statistics
#
# Date created: 19th June 2018
#
# Language:    Python 3
#
# Overview:    
#
# Details:
#
# Usage: to be imported
#
#############################################################

import numpy as np
import scipy as sc

def better_trace (A,N):
    trace = 0
    for n in range(0,N) : trace += A[n,n]
    return trace
    
# Distance between density matrices
#
# Compute the distance between A and
# B using some distance specified by
# type.
#
# A metric d is always obtained from
# a norm using
#
#     d(A,B) = ||A - B||
#
# 1) If type is 'operator', compute the
#    distance based on the operator norm:
#
#    The operator norm of A is
#
#        ||A|| = supp |Ax|
#               |x|=1
#
# 2) If type is 'trace', compute the
#    distance based on the Hilbert-Schmidt
#    norm:
#
#    The Hilbert-Schmidt norm of A is:
#
#        ||A|| = sqrt[ tr(A^ A) ]
#
# 3) If type is 'fidelity', compute the
#    fidelity distance is defined as
#    follows
#
#        F(A,B) = tr[ sqrt(sqrt(A) B sqrt(A)) ]
#        d(A.B) = arccos[F(A,B)]
#
def distance_op(A,B):
    distance = np.linalg.norm(A - B, 2)
    return distance

def distance_trace(A,B):
    distance = np.linalg.norm(A - B,'fro')
    return distance
    
def distance_fid(A,B):
    values_A,vectors_A = np.linalg.eig(A)
    B_new = np.matmul(np.matmul(np.linalg.inv(vectors_A), B), vectors_A)
    A_new = np.diag(np.sqrt(values_A))
    D = np.asmatrix(A_new) * np.asmatrix(B_new) * np.asmatrix(A_new)
    values_D = np.linalg.eigvals(D)
    #fidelity = np.sum(np.linalg.eigvals(D))
    fidelity = np.sum(np.sqrt(values_D))

    C = sc.linalg.sqrtm(A) 
    fidelity1 = np.matrix.trace(sc.linalg.sqrtm(C * B * C))
    print(fidelity1-fidelity)
    exit(1)
    distance = np.arccos(fidelity).real
    return distance

# Use this one if you know the eigenvectors and eigenvalues of
# A already
def distance_fid_2(vectors_A,values_A,B):
    B_new = np.linalg.inv(np.matmul(np.matmul(vectors_A, B), vectors_A))
    A_new = np.diag(np.sqrt(values_A))
    D = np.asmatrix(A_new) * np.asmatrix(B_new) * np.asmatrix(A_new)
    values_D = np.linalg.eigvals(D)
    #fidelity = np.sum(np.linalg.eigvals(D))
    fidelity = np.sum(np.sqrt(values_D))

    #C = sc.linalg.sqrtm(A) 
    #fidelity1 = np.matrix.trace(sc.linalg.sqrtm(C * B * C))
    #print(fidelity1-fidelity)
    #exit(1)
    distance = np.arccos(fidelity).real
    return distance

File: cpp/enm-test/stats.cpp
/************************************************************
* Title: Summary statistics
*
* Date created: 19th June 2018
*
* Language:  Cpp
*
* Overview:    
*
* Details: source
*
* Usage:
*
************************************************************/

#include "stats.h"

// Distance using the operator norm
//
// Here, this is computed by finding
// the largest singular value, but
// I think it might be more complicated
// than that (as in, there are more
// variants of the norm to consider)
//
double distance_op(MatrixXc A, MatrixXc B) {
  Eigen::JacobiSVD<MatrixXc> svd(A-B, Eigen::ComputeThinU | Eigen::ComputeThinV);
  // The first singular value is the largest
  double distance = svd.singularValues()[0];
  return distance;
}

// Distance using the Frobenius norm
//
// This uses a library function from
// Eigen which computes the Frobenius
// (aka Hilbert-Schmidt) norm by
// default
//
double distance_trace(MatrixXc A, MatrixXc B){
  double distance = (A - B).norm();
  return distance;
}

// Fidelity distance
//
// The fidelity distance d is
// defined as follows:
//
//  F(A,B) = tr[ sqrt(sqrt(A) B sqrt(A)) ]
//  d(A.B) = arccos[F(A,B)]
//
// This is computed using a trick that
// involves diagonalising A to ease the
// sqrt function. This originated as a
// python optimisation -- it might not
// be necessary here.
double distance_fid(const MatrixXc A, const MatrixXc B) {

  // Obtain eigenvalues of A
  Eigen::SelfAdjointEigenSolver<MatrixXc> eigenA(A);
  if(eigenA.info() != Eigen::Success) abort();
  MatrixXc vectors_A = eigenA.eigenvectors();
  MatrixXc B_new = vectors_A.inverse() * B * vectors_A;
  // There's a bug hiding in here. The sqrt functions fails
  // if the input is a very small negative number. This needs to
  // be worked around somehow. I've fixed it by taking the
  // absolute value of the eigenvalue before sqrting. 
  MatrixXc A_new(2,2); A_new << std::sqrt(std::abs(eigenA.eigenvalues()[0])),0,
			 0, std::sqrt(std::abs(eigenA.eigenvalues()[1]));
  MatrixXc D = A_new * B_new * A_new;
  Eigen::SelfAdjointEigenSolver<MatrixXc> eigenD(D);
  if(eigenD.info() != Eigen::Success) abort();
  double fidelity = std::sqrt(std::abs(eigenD.eigenvalues()[0]))
    + std::sqrt(std::abs(eigenD.eigenvalues()[0])); 
  double distance = std::acos(fidelity);
  
  return distance;

}

// Fidelity distance 2
//
// This is a more direct implementation of the
// fidelity distance, directly from the definition.
//
double distance_fid_2(const MatrixXc A, const MatrixXc B) {
  Eigen::SelfAdjointEigenSolver<MatrixXc> eigenA(A);
  MatrixXc C = eigenA.operatorSqrt();
  Eigen::SelfAdjointEigenSolver<MatrixXc> eigenCBC(C * B * C);
  MatrixXc D = eigenCBC.operatorSqrt();
  double fidelity = std::real(D.trace());
  double distance = std::acos(fidelity);
  return distance;
}
